### ISSUE TITLE
asyncio, resources, errors, locks, etc.

### ISSUE THREAD

**Issue by robnagler** [radiasoft/sirepo#2467]
Quoted from https://github.com/radiasoft/sirepo/pull/2444#discussion_r416960700:

> @robnagler and I have had a couple calls today to discuss this. I'm just putting down some notes from our discussions.
> 
> We are in agreement that we need to wrap a tornado lock/event/queue/?? so we can ask if we will await and raise awaited when we do. How exactly to implement this is still unclear.
> 
> run_dir_mutex, agent_start_lock, and cpu_slot_q_lock are all doing the same thing. They are guarding a resource that only one thing should be accessing at a time. #2443 arose because we had multiple ops for the same driver all trying to acquire a cpu slot with no coordination.
>
> How we go about solving this is still unclear.
>
> One question we don't have an answer for yet:
> After an op waits on a mutex should it acquire the resource it is after and then raise awaited? Or, should it raise awaited immediately and try to get the resource the next time through the flow? Pros for getting the resource immediately are it follows other patterns (ex waiting on an item from a queue) and means the first thing woken up from the mutex will be the first to get the resource (keeps ordering). The downside to this is that we may grab resources "out of order" which could lead to deadlock. For example, a cpu_slot could be acquired but the websocket is no longer alive (can't really happen because all ops are destroy in this case but it illustrates the idea). In addition we may cycle indefinitely (or until max awaited is met) trying to get a resource.

**Comment by robnagler** [comment]
I wrote [AsyncPython](https://github.com/radiasoft/pykern/wiki/AsynchPython), and it was unfinished

We have had bugs, like #2346 and #2416 and in the past #1082. Deadocks and bugs related to resource management are not new.

We have tried to figure out how to program asynchronously with cooperative multitasking. It's really hard, which is why people write trio, curio, etc. I don't think they fix the problem. I haven't seen a big program that manages resources over long periods of time in Python.

When I start reading about the problem of resources and concurrency, I always end up at shared nothing architectures such as [Erlang](http://www.diva-portal.org/smash/get/diva2:9492/FULLTEXT01.pdf) and [Tandem](https://www.hpl.hp.com/techreports/tandem/TR-85.7.pdf). We have create a shared everything architecture, and we run into bug after bug after bug.

Fortunately, the problem is relatively simple. We can keep it all in our heads (mostly), and we can fix bug after bug while writing tests.

I'm writing this here, because I want to document this here, and keep it open for discussion right now so @e-carlin and I can discuss if there are alternative ways to structure the code to avoid the state problems we continually have.

**Comment by robnagler** [comment]
I'm trying to find examples of how people solve problems with asyncio, and I can't. It's really frustrating. There definitely is confusion out there about cooperative multitasking.

I'm too stupid for AsyncIO](https://whatisjasongoldstein.com/writing/im-too-stupid-for-asyncio/) is the classic confusion about dealing with event loops. This could happen in Javascript or any other single threaded async framework.

People tried to help the author understand [Understanding Asyncio
 ](https://medium.com/@pgjones/understanding-asyncio-a6592a517def) which explains how to write the code in [Quart](https://pgjones.gitlab.io/quart/). The problem itself is so trivial as to be uninteresting:

```py
    futures = [asyncio.ensure_future(fetch(url)) for url in urls]
    output = ""
    for future in futures:
        try:
            resp = await future
        except Exception as error:
            output += "<p><b>Error:</b> %s<p>" % error
            continue
        else:
            output += "<p>%s loaded with a status %s and a length of %s</p>" % (
                resp["url"], resp["resp"].status, len(resp["html"])
            )
    return output
```

Are you kidding me? Embedding `<p>` with `%` strings in asyncio. Just take this:

```py
            output += "<p><b>Error:</b> %s<p>" % error
```

How the heck would you know which url was in error? I know these are examples, but these are BAD examples. And, they don't address issues related to resource management.

[Another article](https://blog.carlmjohnson.net/post/2017/asyncio-link/) explains (without explaining) that Go is better than Python. That's not hard. Rob Pike knows a thing or two about distributed systems. I'll read some Go code now.

**Comment by robnagler** [comment]
Go, of course, is "Do not communicate by sharing memory; instead, share memory by communicating." That we knew. It has other primitives; Rob Pike doesn't recommend them except for special circumstances.

Go uses `select` for concurrent disjunctions. Python uses the event loop for this. I think that's a key issue.

References:
* https://blog.golang.org/pipelines - discusses cancellation
* https://talks.golang.org/2012/concurrency.slide#50 - basic concurrency
* https://talks.golang.org/2013/advconc.slide#1 - advanced concurrency
* https://talks.golang.org/2015/simplicity-is-complicated.slide#1 - why goroutines can't be canceled

**Comment by robnagler** [comment]
Fix 1844f15b3a038c66343f3d27f2f0bf2ee80eedd3

I think we need to do more, but I don't think we should do anything now.
